### PR TITLE
Sensor Observation Service

### LINKED ISSUE #1474: Implement SOS (Sensor Observation Service)
This is a requirement for NEII Viewer (https://github.com/TerriaJS/neii-viewer/issues/2).

The standard: http://www.opengeospatial.org/standards/sos

The target endpoint: http://www.bom.gov.au/waterdata/services?service=SOS&version=2.0&request=GetCapabilities

### PR DESCRIPTION
**Pull request by RacingTadpole**
Fixes #1474.
Test this with http://localhost:3003/#build/terriajs/test/init/sos.json .

Also fixes a few other things -
- improves display of undefined data in a chart preview.
- removes knowledge of TableStructure from FeatureInfoSection.
- passes FeatureInfoSection the real pointer to the Feature, not to a copy. (We may be able to remove Feature.js altogether, but I'll save that for another day.)
- teaches charts about non-csv data formats via `CatalogItem.loadIntoTableStructure`
- improves `tableStyle.scaleByValue` for missing data (scaling them as if 0, not to middle of range)
- adds some comments.

### CODE REVIEW

**Comment by kring** [comment]
In `Water Course Levels`, clicking a point and then clicking the chart's Expand button shows an error:
image

**Comment by kring** [comment]
`Storage Levels` has really big points when displaying Month Average.  @RacingTadpole said it's probably a data problem, judging by the large negative number in the legend.

image

**Comment by kring** [comment]
I haven't seen "Real time" do anything other than show this:
image

**Comment by kring** [comment]
Charts are broken in `BOM SOS Yearly options`:

image

(there's an exception when the feature info loads, too)

File: lib/Models/Feature.js [hunk 1]
```diff
@@ -74,6 +79,31 @@ Feature.fromEntity = function(entity) {
...
+ * @return {Feature} The corresponding feature.
+ */
+Feature.fromEntityCollectionOrEntity = function(entity) {
+    // If this entity is part of a collection, get the feature with this id from that collection.
+    function getHasMatchingIdFunction(entityId) {
+        return function(member) { return member.id === entityId; };
+    }
+    var feature;
+    if (entity.entityCollection) {
+        var matches = entity.entityCollection.values.filter(getHasMatchingIdFunction(entity.id));
```
Comment (kring) [inline lib/Models/Feature.js]: `EntityCollection` has a `getById` function that I think would work here.

File: lib/Models/SensorObservationServiceCatalogItem.js [hunk 2]
```diff
@@ -0,0 +1,1027 @@
...
+var raiseErrorToUser = require('./raiseErrorToUser');
+var TableCatalogItem = require('./TableCatalogItem');
+var TableColumn = require('../Map/TableColumn');
+var TableStructure = require('../Map/TableStructure');
+var TerriaError = require('../Core/TerriaError');
+var VariableConcept = require('../Map/VariableConcept');
+var xml2json = require('../ThirdParty/xml2json');
+
+/**
+ * @alias SensorObservationServiceCatalogItem
```
Comment (kring) [inline lib/Models/SensorObservationServiceCatalogItem.js:41]: Could use an actual description here.

File: wwwroot/test/init/sos.json [hunk 3]
```diff
@@ -0,0 +1,834 @@
...
+                            "identifier": "http://bom.gov.au/waterdata/services/parameters/Water Course Level",
+                            "title": "Water Course Level"
+                        },
+                        {
+                            "identifier": "http://bom.gov.au/waterdata/services/parameters/Water Course Discharge",
+                            "title": "Water Course Discharge"
+                        }
+                    ]
+                }
+                // {
```
Comment (kring) [inline wwwroot/test/init/sos.json]: Do you still want this commented-out stuff in here?

**Comment by kring** [comment]
This is pretty cool!  The code looks good and the functionality works well other than the problems I mentioned above (some of which maybe we can't fix, but we should try to handle them a bit more gracefully anyway).

**Comment by RacingTadpole** [comment]
OK, I think I've addressed all your comments @kring , plus added units to the charts :)
(And in the commit below I've also fixed the big negative number by specifying a `minDisplayValue` of 0.)

**Pull request merged**